Tiltfile

```diff
@@ -1,31 +1,105 @@
+
+    count = int(
+      str(
+        local("lsof -n -i :%s | grep LISTEN | wc -l | tr -d ' '" % default_kibana_port, quiet=True, echo_off=True)
+      ).strip()
+    )
+
+    if count <= 0:
+      printup("No service found on port " + str(default_kibana_port))
+      return None
+
+    is_docker_kibana = str(
+      local(
+        """curl --fail -s -w '%header{kbn-name}'""" + " http://{}:{}".format(default_kibana_host, default_kibana_port),
+        quiet=True,
+        echo_off=True
+      )
+    ).strip() == "kibana"
+
+    if (is_docker_kibana):
+      printup("Running Kibana container found on port " + str(default_kibana_port))
+      return None
+
+
+    kibana_redirect_url = str(
+      local(
+        "curl --fail -s -I -o /dev/null -w '%{redirect_url}' " + 'http://{}:{}'.format(default_kibana_host, default_kibana_port) + ' || [ $? -eq 7 ]',
+        quiet=True,
+        echo_off=True
+      )
+    ).strip().replace("http://", "")
+
+    return kibana_redirect_url
+
 custom_build(
   'apm-server',
   'DOCKER_BUILDKIT=1 docker build -t $EXPECTED_REF -f packaging/docker/Dockerfile .',
   deps = ['.'],
   ignore = ['**/*_test.go', 'tools/**', 'systemtest/**', 'docs/**'],
 )
 
+k8s_yaml(kustomize('testing/infra/k8s/overlays/local'))
+
+k8s_kind('ApmServer', image_json_path='{.spec.image}')
+k8s_kind('Elasticsearch')
+
+
+# Conditionally set up Kibana resource
+discovered_kibana_url = discover_kibana()
+
+if discovered_kibana_url and config.tilt_subcommand == "up":
+  printup("Waiting until local Kibana is ready at {}".format(discovered_kibana_url))
+  
+  # get only the path from the URL so we can pass it separately to APM Server
+  url = discovered_kibana_url
+  if "://" in discovered_kibana_url:
+    url = url.split("://")[1]
+  kibana_path = url.split("/", 1)[1] if "/" in url else ""
+
+  local_resource('kibana', cmd=[ os.path.join(script_dir, 'wait_for_kibana.sh'), "http://admin:changeme@{}".format(discovered_kibana_url)], deps=[], resource_deps=[ 'init_kibana_system_user' ])
+else:
+  printup("Starting Kibana as a container")
+
+  k8s_kind('Kibana')
+  k8s_resource('kibana', port_forwards=default_kibana_port, resource_deps=['elasticsearch'])
+    
 # Build and install the APM integration package whenever source under
 # "apmpackage" changes.
-script_dir = os.path.join(config.main_dir, 'script')
 run_with_go_ver = os.path.join(script_dir, 'run_with_go_ver')
+
+local_resource(
+  "init_kibana_system_user",
+  cmd = """curl -s -X POST -H "Content-Type: application/json" -d '{"username":"kibana_system_user", "password":"changeme", "roles": [ "kibana_system"]}' http://admin:changeme@localhost:9200/_security/user/kibana_system_user""",
+  resource_deps=["elasticsearch"]
+)
+
 local_resource(
   'apmpackage',
   cmd = [os.path.join(script_dir, 'run_with_go_ver'), 'go', 'run', './cmd/runapm -init'],
   dir = 'systemtest',
   deps = ['apmpackage'],
-  resource_deps=['kibana'],
+  resource_deps=['init_kibana_system_user'],
```

Comment:
Should this be depending on "kibana" still? The kibana local_resource depends on init_kibana_system_user.

Reply:
Yes, I've changed this back to `kibana` because it was no longer necessary anyway. I can't remember why I added it here anyway :)